### PR TITLE
c++ replay: fast and reliable merge events

### PR DESCRIPTION
**Pull request by deanlee**
fast and reliable merge events with std::vector. sort events by mono_time and which, the time for merge and sort 5 segments is less than 4ms.

QMultiMap is not suitable for events. it can't guarantee that the events are sent in the correct order without duplicates （mono_time is not UUID), it also takes up more memory and  merging  time. QMultipMap::lower_bound  will also cause the last event to be sent again.

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Nice, once we get these bugs fixed we can delete unlogger. Another pretty bad bug is that replay doesn't work for routes with less segments that replay's window (i.e. only first segment uploaded).

**Comment by deanlee** [comment]
that bug will be fixed in #22203

File: selfdrive/ui/replay/replay.cc [hunk 1]
```diff
@@ -63,13 +63,18 @@ void Replay::mergeEvents() {
   const int end_idx = std::min(current_segment + FORWARD_SEGS, log_paths.size());
 
   // merge logs
-  QMultiMap<uint64_t, Event *> *new_events = new QMultiMap<uint64_t, Event *>();
+  std::vector<Event *> *new_events = new std::vector<Event *>();
   std::unordered_map<uint32_t, EncodeIdx> *new_eidx = new std::unordered_map<uint32_t, EncodeIdx>[MAX_CAMERAS];
   for (int i = start_idx; i <= end_idx; ++i) {
     if (auto it = lrs.find(i); it != lrs.end()) {
```
Comment (deanlee) [inline selfdrive/ui/replay/replay.cc:69]: here is a historic bug. LogerrReader may be downloading and updating events in another thread at this point. that will cause invalid data to be merged, and in the worst case, the program may crash.

Comment (deanlee) [reply selfdrive/ui/replay/replay.cc:69]: fixed in ##22203

**Pull request closed without merging**